components/nimbus/src/behavior.rs

```diff
@@ -198,29 +201,53 @@ impl MultiIntervalCounter {
     }
 }
 
-type EventId = String;
-
-struct EventStore {
-    events: HashMap<EventId, MultiIntervalCounter>,
+#[derive(Default, Serialize, Deserialize, Debug)]
+pub struct EventStore {
+    events: HashMap<String, MultiIntervalCounter>,
 }
 
 impl EventStore {
-    pub fn new(events: Vec<(EventId, MultiIntervalCounter)>) -> Self {
+    pub fn new() -> Self {
+        Self {
+            events: HashMap::<String, MultiIntervalCounter>::new(),
+        }
+    }
+
+    pub fn from_vec(events: Vec<(String, MultiIntervalCounter)>) -> Self {
         Self {
             events: HashMap::from_iter(events.into_iter()),
         }
     }
 
-    pub fn from(events: HashMap<EventId, MultiIntervalCounter>) -> Self {
+    pub fn from(events: HashMap<String, MultiIntervalCounter>) -> Self {
         Self { events }
     }
 
-    pub fn record_event(&mut self, event_id: EventId, now: Option<DateTime<Utc>>) -> Result<()> {
+    pub fn from_persisted_data(db: &Database) -> Result<Self> {
+        let reader = db.read()?;
+        let events = db
+            .get_store(StoreId::EventCounts)
+            .collect_all::<(String, MultiIntervalCounter), _>(&reader)?;
+        Ok(EventStore::from_vec(events))
+    }
+
+    pub fn record_event(&mut self, event_id: String, now: Option<DateTime<Utc>>) -> Result<()> {
         let now = now.unwrap_or_else(Utc::now);
         let counter = self.events.get_mut(&event_id).unwrap();
         counter.maybe_advance(now)?;
         counter.increment()
     }
+
+    pub fn persist_data(&self, db: &Database) -> Result<()> {
+        let mut writer = db.write()?;
+        let data: Vec<(&String, &MultiIntervalCounter)> = (&self.events).iter().collect();
+        for (k, v) in data {
+            let d: (String, MultiIntervalCounter) = (k.clone(), v.clone());
+            db.get_store(StoreId::EventCounts).put(&mut writer, k, &d)?;
+        }
```

Comment:
nit: we can avoid the vec allocation by using `try_for_each` here
```suggestion
        self.events.iter().try_for_each(|(key, value)| {
            db.get_store(StoreId::EventCounts).put(&mut writer, key, &(key.clone(), value.clone()))
        })?;
```